File: lana/src/main/scala/com/financialforce/lana/Context.scala
/*
 * Copyright (c) 2020 FinancialForce.com, inc. All rights reserved.
 */
package com.financialforce.lana

import com.financialforce.lana.commands.{LoadLogFile, ShowLogFile}
import com.financialforce.lana.runtime.vscode.ExtensionContext
import com.financialforce.lana.runtime.vscode.workspace.workspaceFolders
import com.financialforce.lana.workspace.VSWorkspace

import scala.scalajs.js

class Context(val context: ExtensionContext, val display: Display, val debugMode: Boolean) {

  val symbolFinder = new SymbolFinder()
  val workspaces: Seq[VSWorkspace] = workspaceFolders.getOrElse(js.Array()) map (VSWorkspace(_, display))
  val namespaces: Seq[String] = Seq()

  register()

  def findSymbol(wsPath: String, symbol: String): Option[String] = {
    try {
      Some(symbolFinder.findSymbol(wsPath, symbol))
    } catch {
      case ex: Exception =>
        display.showErrorMessage(ex.getMessage)
        None
    }
  }

  private def register(): Unit = {
    LoadLogFile(this)
    ShowLogFile(this)
  }
}

object Context {
  def apply(context: ExtensionContext, display: Display, debugMode: Boolean): Context = {
    new Context(context, display, debugMode)
  }
}

File: src/Context.ts
/*
 * Copyright (c) 2020 FinancialForce.com, inc. All rights reserved.
 */

import { ExtensionContext, workspace } from "vscode";
import { LoadLogFile } from "./commands/LoadLogFile";
import { ShowLogFile } from "./commands/ShowLogFile";
import { Display } from "./Display";
import { SymbolFinder } from "./SymbolFinder";
import { VSWorkspace } from "./workspace/VSWorkspace";

export class Context {
  symbolFinder = new SymbolFinder();
  context: ExtensionContext;
  display: Display;
  workspaces: VSWorkspace[] = [];
  namespaces: string[] = [];

  constructor(context: ExtensionContext, display: Display) {
    this.context = context;
    this.display = display;

    if (workspace.workspaceFolders) {
      this.workspaces = workspace.workspaceFolders.map((folder) => {
        return new VSWorkspace(folder);
      });
    }

    LoadLogFile.apply(this);
    ShowLogFile.apply(this);
  }

  findSymbol(wsPath: string, symbol: string): string | null {
    try {
      return this.symbolFinder.findSymbol(wsPath, symbol);
    } catch (err) {
      this.display.showErrorMessage(err.message);
      return null;
    }
  }
}
